### PR TITLE
[Test] Fix flaky test_cli.py

### PR DESCRIPTION
**Pull request by rkooo567**
Signed-off-by: SangBin Cho <rkooo567@gmail.com>

## Why are these changes needed?

Looks like this test has been innately flaky because of shutdown order upon `ray stop`. For example, monitor.py sometimes exits ungracefully because GCS exits before monitor, which makes it raise an exception and fail (which makes this test flaky). This PR fixes the flakiness and also the general robustness of shutdown by shutting down all processes first -> shutting down GCS after that upon `ray stop`.

As a side effect, it can take up to 2X of grace_period to terminate all processes (because we kill all other processes first -> kill GCS). 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Okay. I ran tests 50 times, and it worked all the times...

**Review by DmitriGekhtman — APPROVED** [review]

File: python/ray/scripts/scripts.py [hunk 1]
```diff
@@ -964,122 +964,163 @@ def start(
...
+
+        Returns:
+            total_procs_found: Total number of processes found from
+                `processes_to_kill` is added.
+            total_procs_stopped: Total number of processes gracefully
+                stopped from `processes_to_kill` is added.
+            procs_not_gracefully_killed: If processes are not killed
+                gracefully, they are added here.
+        """
+        process_infos = []
```
Comment (rkooo567) [inline python/ray/scripts/scripts.py:998]: All the code within this method hasn't been changed. It is same as before.

File: python/ray/scripts/scripts.py [hunk 2]
```diff
@@ -964,122 +964,163 @@ def start(
...
+
+        # For processes that are not killed within the grace period,
+        # we send force termination signals.
+        for proc in alive:
+            proc.kill()
+        # Wait a little bit to make sure processes are killed forcefully.
+        psutil.wait_procs(alive, timeout=2)
+        return total_found, total_stopped, alive
+
+    # GCS should exit after all other processes exit.
```
Comment (rkooo567) [inline python/ray/scripts/scripts.py:1086]: From here, it is new code (we aggregate the result and print result only once)

File: python/ray/tests/test_cli.py [hunk 3]
```diff
@@ -460,84 +471,99 @@ def test_ray_start_block_and_stop(configure_lang, monkeypatch, tmp_path):
...
-    # termination by signals
-    if not head_proc.is_alive():
-        # NOTE(rickyyx): call recv() here is safe since the process
-        # is guaranteed to be terminated.
-        if head_proc.exitcode == 1:
-            assert False, (
-                "ray start --head --block is failed "
-                "due to unexpected component failures."
+    try:
+        # Run
```
Comment (rkooo567) [inline python/ray/tests/test_cli.py:475]: this doesn't change a test as well. It just improves the cleaning logic.

**Comment by rkooo567** [comment]
> As a side effect, it can take up to 2X of grace_period to terminate all processes (because we kill all other processes first -> kill GCS).

Q: is it the desired behavior? Should we document this? Or is there any better solution?

File: python/ray/scripts/scripts.py [hunk 4]
```diff
@@ -964,122 +964,163 @@ def start(
...
-                    proc.terminate()
-
-                if force:
+                process_infos.append((proc, proc.name(), proc.cmdline()))
+            except psutil.Error:
+                pass
+
+        stopped = []
+        for keyword, filter_by_cmd in processes_to_kill:
+            if filter_by_cmd and is_linux and len(keyword) > 15:
```
Comment (rickyyx) [inline python/ray/scripts/scripts.py:1007]: do u know why 15?

**Review by rickyyx — APPROVED** [review]
> As a side effect, it can take up to 2X of grace_period to terminate all processes (because we kill all other processes first -> kill GCS).

Wonder if we should either document this / or enforce an overall `grace_period`, e.g. a timer that fires that forced kills everything if graceful shutdown of non-gcs and gcs exceeds the graceful period.

File: python/ray/scripts/scripts.py (hunk 4 above)
Comment (rkooo567) [reply python/ray/scripts/scripts.py:1007]: not sure why... this was the code that was there before

**Comment by rkooo567** [comment]
Hmm what about we just double the default and wait grace_period / 2 for each step?

**Comment by rickyyx** [comment]
> Hmm what about we just double the default and wait grace_period / 2 for each step?

Will also work - but changes the `grace_period` definition a bit? I guess with some documentation, it's fine.

**Pull request merged**